File: mhike-school-web/lib/services/notificationPreferences.ts
import { apiGet, apiPatch } from "@/lib/api";

export interface NotificationPreferences {
    id: number;
    school_id: number;
    user_id: number;
    attendance_alerts_enabled: boolean;
    absence_notifications_enabled: boolean;
    persistent_absence_alerts_enabled: boolean;
    safeguarding_alerts_enabled: boolean;
    email_enabled: boolean;
    push_enabled: boolean;
    sms_enabled: boolean;
}

export interface UpdateNotificationPreferencesPayload {
    attendance_alerts_enabled?: boolean;
    absence_notifications_enabled?: boolean;
    persistent_absence_alerts_enabled?: boolean;
    safeguarding_alerts_enabled?: boolean;
    email_enabled?: boolean;
    push_enabled?: boolean;
    sms_enabled?: boolean;
}

export function getNotificationPreferences(
    token?: string,
): Promise<NotificationPreferences> {
    return apiGet<NotificationPreferences>(
        "/notification-preferences/me",
        token,
    );
}

export function updateNotificationPreferences(
    payload: UpdateNotificationPreferencesPayload,
    token?: string,
): Promise<NotificationPreferences> {
    return apiPatch<NotificationPreferences>(
        "/notification-preferences/me",
        payload,
        token,
    );
}

File: app/models/notification_preferences.py
from __future__ import annotations

from sqlalchemy import Boolean, ForeignKey, Integer
from sqlalchemy.orm import Mapped, mapped_column

from app.db.base import Base


class NotificationPreference(Base):
    __tablename__ = "notification_preferences"

    id: Mapped[int] = mapped_column(
        Integer,
        primary_key=True,
        index=True,
    )

    school_id: Mapped[int] = mapped_column(
        ForeignKey("schools.id"),
        nullable=False,
        index=True,
    )

    user_id: Mapped[int] = mapped_column(
        ForeignKey("users.id"),
        nullable=False,
        index=True,
    )

    attendance_alerts_enabled: Mapped[bool] = mapped_column(
        Boolean,
        default=True,
        nullable=False,
    )

    absence_notifications_enabled: Mapped[bool] = mapped_column(
        Boolean,
        default=True,
        nullable=False,
    )

    persistent_absence_alerts_enabled: Mapped[bool] = mapped_column(
        Boolean,
        default=True,
        nullable=False,
    )

    safeguarding_alerts_enabled: Mapped[bool] = mapped_column(
        Boolean,
        default=True,
        nullable=False,
    )

    email_enabled: Mapped[bool] = mapped_column(
        Boolean,
        default=True,
        nullable=False,
    )

    push_enabled: Mapped[bool] = mapped_column(
        Boolean,
        default=False,
        nullable=False,
    )

    sms_enabled: Mapped[bool] = mapped_column(
        Boolean,
        default=False,
        nullable=False,
    )
